### ISSUE TITLE
[BUG][Python] Asyncio cleanup fails when exceptions are raised

### ISSUE THREAD

**Issue by TBBle** [OpenAPITools/openapi-generator#5538]
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [ ] ~Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?~
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] ~[Optional] Bounty to sponsor the fix~

##### Description

When an exception occurs in python `--library asyncio` generated code, it fails to clean up correctly due to use of the [`__del__` magic method to clean up the aiohttp `ClientSession` object](https://github.com/fabianvf/openapi-generator/blob/0366e946ba9d1a30a777cf085f24cffd7f404211/modules/openapi-generator/src/main/resources/python/asyncio/rest.mustache#L80).

This is due to using `RESTClientObject.__del__` to schedule `aiohttp.ClientSession.close` with `asyncio.ensure_future`. `__del__` doesn't run until the exception is discarded, and in this repro case, that happens outside the asyncio event loop, and hence raises an exception itself.

##### openapi-generator version

4.3.0-SNAPSHOT, pulled as follows:

```
PS C:\Users\paulh> docker pull openapitools/openapi-generator-cli
Using default tag: latest
latest: Pulling from openapitools/openapi-generator-cli
709515475419: Already exists
38a1c0aaa6fd: Already exists
cd134db5e982: Already exists
044751432162: Pull complete
d41fd233bfb5: Pull complete
b9e8842b3734: Pull complete
Digest: sha256:76b6fc8b94789382e1d91319b787545ec5b7f29c3709c2c14bc47eab68d3a871
Status: Downloaded newer image for openapitools/openapi-generator-cli:latest
docker.io/openapitools/openapi-generator-cli:latest
```
```
PS C:\Users\paulh> docker image ls openapitools/openapi-generator-cli
REPOSITORY                           TAG                 IMAGE ID            CREATED             SIZE
openapitools/openapi-generator-cli   latest              e74d7eb8269d        29 minutes ago      135MB
```

I don't believe this is a recent regression, the code hasn't been touched since it was introduced in #107 in 2018.

##### OpenAPI declaration file content or url

<https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/2_0/petstore.yaml>

##### Command line used for generation

```
docker run --rm -v ${PWD}:/local openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/2_0/petstore.yaml -g python --library asyncio -o /local/out/python
```
##### Steps to reproduce

After code-gen, create and run the following Python script:
```python
import asyncio
import sys

sys.path.append("out/python")
import openapi_client
from openapi_client.rest import ApiException

async def main():
    configuration = openapi_client.Configuration()

    with openapi_client.ApiClient(configuration) as api_client:

        api_instance = openapi_client.PetApi(api_client)

        try:
            # Trigger a 404 search result
            result = await api_instance.get_pet_by_id("nosuchpet")
        except ApiException as e:
            print(f"get_pet_by_id failed: {e.reason}")
            assert e.status == 404
            raise e


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except ApiException as e:
        assert e.status == 404
        pass
```

###### Actual output

```
get_pet_by_id failed: Not Found
Exception ignored in: <function RESTClientObject.__del__ at 0x00000260D05EC1E0>
Traceback (most recent call last):
  File "out/python\openapi_client\rest.py", line 89, in __del__
    asyncio.ensure_future(self.pool_manager.close())
  File "C:\Program Files\Python37\lib\asyncio\tasks.py", line 580, in ensure_future
    loop = events.get_event_loop()
  File "C:\Program Files\Python37\lib\asyncio\events.py", line 644, in get_event_loop
    % threading.current_thread().name)
RuntimeError: There is no current event loop in thread 'MainThread'.
sys:1: RuntimeWarning: coroutine 'ClientSession.close' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x00000260D05F12B0>
```

###### Expected output

```
get_pet_by_id failed: Not Found
```

##### Related issues/PRs

None I'm aware of.

###### Workaround

A monkey-patch to remove `__del__` and a `finally` block to directly call `aiohttp.ClientSession.close` within the asyncio context fixes the issue from the user side:

```
import asyncio
import sys

sys.path.append("out/python")
import openapi_client
from openapi_client.rest import ApiException

async def main():
    configuration = openapi_client.Configuration()

    with openapi_client.ApiClient(configuration) as api_client:

        api_instance = openapi_client.PetApi(api_client)

        try:
            # Trigger a 404 search result
            result = await api_instance.get_pet_by_id("nosuchpet")
        except ApiException as e:
            print(f"get_pet_by_id failed: {e.reason}")
            assert e.status == 404
            raise e
        finally:
            await api_client.rest_client.pool_manager.close()

def monkey():
    from openapi_client.rest import RESTClientObject

    del RESTClientObject.__del__


monkey()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except ApiException as e:
        assert e.status == 404
        pass
```

----

My ideal solution would be if `ApiClient` in the asyncio config was an [asynchronous context manager](https://www.python.org/dev/peps/pep-0492/#asynchronous-context-managers-and-async-with), which would mean using it as:

```
async with openapi_client.ApiClient(configuration) as api_client:
```

Then its `__aexit__` code could call an `async def close()` on `RESTClientObject`, which would close its `aiohttp.ClientSession`.

However, the _tornado_ and _urllib_ implementations of `RESTClientObject` do not have any particular need to be notified on `ApiClient.close`, so this would be introducing an asyncio-specific path.

**Comment by TBBle** [comment]
As a side-note (probably needs to be a separate bug report, putting it here so I don't forget about it), `async_req=True` doesn't work with the _asyncio_ library, as internally it passes the `__call_api` method to `multiprocessing.apply_async`, but with _asyncio_, `__call_api` is a couroutine and so it is never awaited. Instead, it looks like this should be using [`concurrent.futures.ThreadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) and [`asyncio.loop.run_in_executor`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).

Repro is pretty simple, changing the repro case from
```python
result = await api_instance.get_pet_by_id("nosuchpet")
```

to
```python
result = api_instance.get_pet_by_id("nosuchpet", async_req=True)
result.wait()
```

triggers the failure output, instead of the same expected output.

```
C:\Program Files\Python37\lib\asyncio\events.py:88: RuntimeWarning: coroutine 'ApiClient.__call_api' was never awaited
  self._context.run(self._callback, *self._args)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Comment by TBBle** [comment]
I haven't tested it yet, but it looks like #5621 fixed this.

**Comment by TBBle** [comment]
Confirmed fixed with latest OpenAPI-Generator Docker image
```
PS > docker pull openapitools/openapi-generator-cli
Using default tag: latest
latest: Pulling from openapitools/openapi-generator-cli
709515475419: Already exists
38a1c0aaa6fd: Already exists
cd134db5e982: Already exists
6e1edfcbca40: Pull complete
6fe4a4ff99ff: Pull complete
b0db1e10806b: Pull complete
Digest: sha256:09ba860ec2290e41b9b813a0dae6a71f7c6b5476c8312791ca875f91d6a031d9
Status: Downloaded newer image for openapitools/openapi-generator-cli:latest
docker.io/openapitools/openapi-generator-cli:latest
```

The same command-line for generation as used above, now reports:

```
PS > cat .\out\python\.openapi-generator\VERSION
4.3.1-SNAPSHOT
```

And the repro case passes, with one minor change due to API change from this fix:

```python
    with openapi_client.ApiClient(configuration) as api_client:
```
is now
```python
    async with openapi_client.ApiClient(configuration) as api_client:
```

I also confirmed this fix in the 4.3.0 release, repeating the reproduction case using Docker image _openapitools/openapi-generator-cli:v4.3.0_.
